doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,321 @@
+kind of Source isn't entirely clear.
+2. `INSERT`s into a table are linearizable, whereas requests from a webhook should not be. We can
+update `TABLE`s to make `INSERT`s optionally linearizable, but this work probably isn't worth it
+given we also have reason 1.
+
+The second alternative is to build this new kind of storage owned Source, which gets installed on a
+cluster and can do more complex operations everytime a request is received. We plan to build
+something like this in the future, but at the moment this would require a large amount of work,
+and we first we want to validate that webhooks are a feature users actually want.
+
+## Future
+
+There is plenty of future work that can be done for Webhooks, but by limiting our scope and
+shipping quickly, we can collect feedback from users to help inform our priorities.
+
+1. A `clusterd` primarily handles requests. This would improve our ability to scale, e.g. 100s of
+QPS per source, and allow us to build the more complex features listed below.
+2. `TRANSFORM USING` a given request. Being able to map a JSON (or Avro, protobuf, etc.) body to
+a relation at the time of receiving a request, would be a win for the user as it's more ergonomic,
+and a win for us to reduce the amount of storage used.
+3. Subsources and batching. Being able map a single event to multiple independent sources, could
+be a big win for the ergonomics of the feature.
+4. User configuration of request limits via `ALTER <source>`, i.e. rate and size.
+5. Consider using some "external ID" that users can use to reference Webhook sources, e.g.
+   `/api/webhook/abcdefg`. This would allow users to drop and re-create sources without needing to
+   update external systems.
+   See <https://github.com/MaterializeInc/materialize/pull/20002#discussion_r1234726035> for more
+   details.
+6. Supporting "recipes" for request validation for common webhook applications. For example,
+   `VALIDATE STRIPE (SECRET ...)` would handle HMAC-ing the body of the request, and matching it
+   against the exact header that Stripe expects.
+
+
+As a _rough sketch_, the north star for this feature in terms of SQL syntax could be:
+
+```
+CREATE SOURCE <name> FROM WEBHOOK
+  -- parse the body as JSON.
+  BODY FORMAT JSON,
+  INCLUDE HEADERS,
+  -- reject the request if this expression returns False.
+  VALIDATE USING (
+    SELECT headers->>'signature' = sha256hash(body)
+  ),
+  -- transform the body and headers into a well formed relation, before persisting it.
+  AS (
+    SELECT
+        body->>'id' as id,
+        body->>'name' as name,
+        body->>'event_type' as event_type,
+        body->>'country' as country,
+        headers->>'X-Application' as application
+  ),
+  -- create a subsource that contains specific info.
+  SUBSOURCE <name> (
+    SELECT
+      body->>'id' as id,
+      body->>'ip_addr' as id_address
+  );
+```
+
+## Open questions
+
```

Comment:
A few open questions for future work. Since we're adding things that are open to the Internet here, I'm thinking about:
 - firewall rules on _our_ side for ensuring that the environment doesn't get slammed by random noise
 - related to the above, ability on _our_ side to suspend the source if it's causing unnecessary load
 - consistency / permanence of DNS records (means some of the things the cloud team has thrown around re: faster provisioning might need to change)
 - observability for the customer on webhook handling (i.e. to answer the question of "I just sent the test webhook from Stripe, why isn't the data showing up")

Reply:
These are great points, I added them all to open questions! I plan to turn each of these into a GitHub issue so we can track them and have pointed discussions.

Reply:
Opened https://github.com/MaterializeInc/materialize/issues/20220 to make sure we evaluate the security risk here, and come up with a plan